sql/planbuilder/create_ddl.go

```diff
@@ -192,60 +200,229 @@ func (b *Builder) buildCreateProcedure(inScope *scope, subQuery string, fullQuer
+				if err != nil || number == 0 {
+					// We use our own error instead
+					err := fmt.Errorf("invalid value '%s' for MySQL error code", string(dc.MysqlErrorCode.Val))
+					b.handleErr(err)
+				}
+				//TODO: implement MySQL error code support
+				err = sql.ErrUnsupportedSyntax.New(ast.String(s))
+				b.handleErr(err)
+			}
+			inScope.proc.AddCondition(plan.NewDeclareCondition(dc.Name, 0, ""))
+		} else if s.Variables != nil {
+			typ, err := types.ColumnTypeToType(&s.Variables.VarType)
+			if err != nil {
+				b.handleErr(err)
+			}
+			for _, v := range s.Variables.Names {
+				varName := strings.ToLower(v.String())
+				param := expression.NewProcedureParam(varName, typ)
+				inScope.proc.AddVar(param)
+				inScope.newColumn(scopeColumn{col: varName, typ: typ, scalar: param})
+			}
+		} else if s.Cursor != nil {
+			inScope.proc.AddCursor(s.Cursor.Name)
+		} else if s.Handler != nil {
+			switch s.Handler.ConditionValues[0].ValueType {
+			case ast.DeclareHandlerCondition_NotFound:
+			case ast.DeclareHandlerCondition_SqlException:
+			default:
+				err := sql.ErrUnsupportedSyntax.New(ast.String(s))
+				b.handleErr(err)
+			}
+			inScope.proc.AddHandler(nil)
+		}
+	case *ast.BeginEndBlock:
+		blockScope := inScope.push()
+		blockScope.initProc()
+		blockScope.proc.AddLabel(s.Label, false)
+		b.validateBlock(blockScope, s.Statements)
+	case *ast.Loop:
+		blockScope := inScope.push()
+		blockScope.initProc()
+		blockScope.proc.AddLabel(s.Label, true)
+		b.validateBlock(blockScope, s.Statements)
+	case *ast.Repeat:
+		blockScope := inScope.push()
+		blockScope.initProc()
+		blockScope.proc.AddLabel(s.Label, true)
+		b.validateBlock(blockScope, s.Statements)
+	case *ast.While:
+		blockScope := inScope.push()
+		blockScope.initProc()
+		blockScope.proc.AddLabel(s.Label, true)
+		b.validateBlock(blockScope, s.Statements)
+	case *ast.IfStatement:
+		for _, cond := range s.Conditions {
+			b.validateBlock(inScope, cond.Statements)
+		}
+		if s.Else != nil {
+			b.validateBlock(inScope, s.Else)
+		}
+	case *ast.Iterate:
+		if exists, isLoop := inScope.proc.HasLabel(s.Label); !exists || !isLoop {
+			err := sql.ErrLoopLabelNotFound.New("ITERATE", s.Label)
+			b.handleErr(err)
+		}
+	case *ast.Signal:
+		if s.ConditionName != "" {
+			signalName := strings.ToLower(s.ConditionName)
+			condition := inScope.proc.GetCondition(signalName)
+			if condition == nil {
+				err := sql.ErrDeclareConditionNotFound.New(signalName)
+				b.handleErr(err)
+			}
+		}
+	case *ast.FetchCursor:
+		if !inScope.proc.HasCursor(s.Name) {
+			b.handleErr(sql.ErrCursorNotFound.New(s.Name))
+		}
+	case *ast.OpenCursor:
+		if !inScope.proc.HasCursor(s.Name) {
+			b.handleErr(sql.ErrCursorNotFound.New(s.Name))
+		}
+	case *ast.CloseCursor:
+		if !inScope.proc.HasCursor(s.Name) {
+			b.handleErr(sql.ErrCursorNotFound.New(s.Name))
+		}
+
+	// limit validation
+	case *ast.Select:
+		if s.Limit != nil {
```

Comment:
the limit/offset in analyzer seems like it does a lot, was this for a specific test that regular analyzer loop doesn't catch?

Reply:
Yep, `create procedure` checks that the limit/offset value is actually a numeric.